### PR TITLE
feat(compile): compile w/dependencies and options

### LINKED ISSUE #952: Allow compilation with dependencies from WASM
### Problem

I can compile Noir program in JavaScript with noir_wasm but I can not use `dep::std` or any other dependency because compiler would not resolve it. 

### Proposed solution

Add the ability to "declare" non-local crate while compiling

### Alternatives considered

Using flat file technique and in-line all dependencies into a single file.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by kobyhallx**
# Related issue(s)

#952 Allow compilation with dependencies from WASM, https://github.com/noir-lang/noir/issues/888#issuecomment-1453388917

Resolves #952 Allow compilation with dependencies from WASM

# Description

Allows to call `compile` with a set of dependencies as a second argument. Those dependencies will be later resolved from [noir-lang/noir-source-resolver](https://github.com/noir-lang/noir-source-resolver)

## Summary of changes

[crates/wasm/src/lib.rs](https://github.com/noir-lang/noir/compare/kh-wasm-compile-w-dependency?expand=1#diff-6ce577db55a8f1aababd2da61a6292df631cbc475f3282c5a58e0dbe58987f07R35) adds the additional argument which will be iterated over to resolves dependencies. It's equivalent to declaring it in project configuration.

## Dependency additions / changes

`noirc_driver` is now required by `wasm`

## Test additions / changes

N/A

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.

## Documentation needs
- [x] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

File: crates/wasm/src/lib.rs [hunk 1]
```diff
@@ -15,21 +17,72 @@ pub struct BuildInfo {
     dirty: &'static str,
 }
 
+#[derive(Serialize, Deserialize)]
+pub struct WASMCompileOptions {
+    // Compile each contract function used within the program
+    #[serde(default = "bool::default")]
```
Comment (jfecher) [inline crates/wasm/src/lib.rs:23]: Is the `= "bool::default"` needed here?

File: crates/wasm/src/lib.rs [hunk 2]
```diff
@@ -15,21 +17,72 @@ pub struct BuildInfo {
...
+
+        <JsValue as JsValueSerdeExt>::from_serde(&collected_compiled_programs).unwrap()
+    } else {
+        let main = driver.main_function();
+        let compiled_program = driver
+            .compile_no_check(&options.compile_options, main)
+            .unwrap_or_else(|_| panic!("Compilation failed"));
+
+        <JsValue as JsValueSerdeExt>::from_serde(&compiled_program).unwrap()
+    }
```
Comment (jfecher) [inline crates/wasm/src/lib.rs:85]: Duplicating this logic across the compile command and the wasm crate is a bit unwieldy. Is there any way we can factor out the common functionality?

File: crates/wasm/src/lib.rs (hunk 1 above)
Comment (kobyhallx) [reply crates/wasm/src/lib.rs:23]: yes, to default if not provided as an option

File: crates/wasm/src/lib.rs (hunk 2 above)
Comment (TomAFrench) [reply crates/wasm/src/lib.rs:85]: I had similar thoughts on this. Currently working on a separate PR to separate compilation from saving/preprocessing the compiled programs which we can then merge into this one.

Comment (kobyhallx) [reply crates/wasm/src/lib.rs:85]: yes, this needs to be extracted and unified across both cli and wasm. Need to put it off for now.

**Review by jfecher — DISMISSED** [review]

**Review by TomAFrench — APPROVED** [review]
I'm not a huge fan of the multiple return value types (single program vs an array of programs) but I don't think it's a blocker.

**Comment by kobyhallx** [comment]
> I'm not a huge fan of the multiple return value types (single program vs an array of programs) but I don't think it's a blocker.

What's the alternative you're thinking about?

**Comment by TomAFrench** [comment]
> What's the alternative you're thinking about?

My preference is to have separated logic for compiling contracts vs single programs which is then reusable here. 
i.e. `nargo compile` determines whether you want a contract or program, call the relevant function which returns the compiled artifacts (either single or vector) and then writes these to disk. We then reuse the middle step in this crate.

> I was going for party with CLI first.

Yep, it matching our implementation in the CLI is why I'm fine with this going through. We can refactor later as we define the compilation code-paths for programs/contracts more.

That said I think that having multiple "return types" is a bit different when writing to the FS rather than passing values back to code.

**Comment by kobyhallx** [comment]
> > What's the alternative you're thinking about?
> 
> My preference is to have separated logic for compiling contracts vs single programs which is then reusable here. i.e. `nargo compile` determines whether you want a contract or program, call the relevant function which returns the compiled artifacts (either single or vector) and then writes these to disk. We then reuse the middle step in this crate.
> 
> > I was going for party with CLI first.
> 
> Yep, it matching our implementation in the CLI is why I'm fine with this going through. We can refactor later as we define the compilation code-paths for programs/contracts more.
> 
> That said I think that having multiple "return types" is a bit different when writing to the FS rather than passing values back to code.

Agreed, this is still WIP to be addressed.

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**